man/mem-monitor.1
.TH MEM-MONITOR 1 "2008-03-19" "sp-memusage"
.SH NAME
mem-monitor - output system memory usage at given intervals
.SH SYNOPSIS
mem-monitor [interval in secs]
.SH DESCRIPTION
\fImem-monitor\fP outputs (one-liner) system memory usage information from
/proc/meminfo at given intervals.  By default the interval is 3 seconds.
.PP
This is obsoleted by \fImem-cpu-monitor\fP binary which can show also
specified processes memory and CPU usage.
.SH EXAMPLE OUTPUT
Example \fImem-monitor\fP output:
.br
	time:           total:  free:   used:   use-%:  status:
.br
	14:06:26        3220376 2986320 234056  7
.br
	14:06:29        3220376 2986304 234072  7
.SH SEE ALSO
.IR proc (1),
.IR mem-cpu-monitor (1)
.SH COPYRIGHT
Copyright (C) 2007,2008 Nokia Corporation.
.PP
This is free software.  You may redistribute copies of it under the
terms of the GNU General Public License v2 included with the software.
There is NO WARRANTY, to the extent permitted by law.


changelog
sp-memusage (1.3.1), 27 Dec 2011:
  * split mem-cpu-plot output to separate graphs for
    each of the processes and the system.
  * improve graph texts and error handling.
  * update manual pages.

sp-memusage (1.3), 1 Nov 2011:
  * mem-cpu-plot: new utility for plotting mem-cpu-monitor output with
    gnuplot.

sp-memusage (1.2.15), 9 Sep 2011:
  * mem-cpu-monitor: apply column coloring also to cgroups.

sp-memusage (1.2.14), 5 Aug 2011:
  * mem-cpu-monitor: fix potential use of uninitialized data.

sp-memusage (1.2.13), 30 Jun 2011:
  * mem-cpu-monitor: fix process mem/cpu change monitoring flags (-c/-m).

sp-memusage (1.2.12), 30 Jun 2011:
  * mem-cpu-monitor: change --cgroup option argument from optional to
    required.

sp-memusage (1.2.11), 21 Jun 2011:
  * mem-smaps-totals: add support for Locked and Anonymous smaps fields.
  * mem-cpu-monitor: add support for monitoring multiple cgroups.

sp-memusage (1.2.10), 17 May 2011:
  * Fix mem-smaps-totals incorrect help output.

sp-memusage (1.2.9), 13 May 2011:
  * Fix mem-cpu-monitor --pid option breakage in 1.2.7.
  * Fix mem-cpu-monitor column coloring.

sp-memusage (1.2.8), 12 May 2011:
  * Fix mem-cpu-monitor interval (-i) option breakage in 1.2.7.

sp-memusage (1.2.7), 10 May 2011:
  * Add Control Group (aka. Cgroup) memory usage monitoring option to
    mem-cpu-monitor

sp-memusage (1.2.6), 14 Apr 2011:
  * mem-cpu-monitor didn't notice when named process starts

sp-memusage (1.2.5), 24 Feb 2011:
  * mem-cpu-monitor log has multiple columns with the same PID

sp-memusage (1.2.4), 10 Jan 2011:
  * sp-memusage-tests packaging fixed.

sp-memusage (1.2.3), 29 Dec 2010:
  * Fix mem-cpu-monitor --no-colors argument. 
  * New script: mem-smaps-totals

sp-memusage (1.2.2), 20 Jul 2010:
  * mem-smaps-private script show PSS and it's possible to give it multiple
    PIDs. 

sp-memusage (1.2.1), 10 Jun 2010:
  * Show correct process name when using the --exec option or when
    the process name is otherwise changes.  

sp-memusage (1.2), 16 Apr 2010:
  * mem-cpu-monitor argument parsing change:
    - setting interval requires use of -i option
    - align names of process option names with system options
      (mem-changes -> mem-change, cpu-changes -> cpu-change)
  * mem-cpu-monitor updates:
    - support for sub-second intervals
    - refactored to use/depend on the new libsp-measure library
    - support for starting monitored process (-exec) and monitoring
      processes starting after mem-cpu-monitor starts (-name).
    - if some resource subset is unavailable, warn about it & exclude
      the resource from further processing, but let program continue
      (e.g. if kernel is lacking CPU frequency scaling)

sp-memusage (1.1.6), 18 Jan 2010:
  * Do not shows private clean data changes in the "changes" column.

sp-memusage (1.1.5), 3 Nov 2009:
  * Support swap in mem-smaps-private and mem-dirty-code-pages scripts.

sp-memusage (1.1.4), 26 Oct 2009:
  * mem-cpu-monitor:
   + mem-cpu-monitor: report CPU speed. 
   + mem-cpu-monitor: implementation of options that make it output info only
     when something changes. 

sp-memusage (1.1.3), 14 Jul 2009:
  * Add a manual page for mem-cpu-monitor.

sp-memusage (1.1.2), 15 Jun 2009:
  * Always have something in BL column, if it's shown.

sp-memusage (1.1.1), 29 May 2009:
  * Partial rewrite of mem-cpu-monitor.  Supports giving multiple pids,
    has colors when running on tty, flushes output when redirected to
    file, /proc reading refactored to common file shared with
    mem-monitor. 

sp-memusage (1.1.0), 28 Apr 2009:
  * mem-cpu-monitor added.

sp-memusage (1.0.12), 13 Nov 2008:
  * SwapCached is now also taken in account while calculating memory
    statistics. 

sp-memusage (1.0.11), 17 Sep 2008:
  * Some minor packaging issues reported by Lintian have been fixed.

sp-memusage (1.0.10), 11 Sep 2008:
  * Mem-monitor-smaps now uses the latest pid returned when there are
    multiple matches.
  * As 'available' is more approriate than 'free' in the context of
    generated reports, use that instead.
  * New option (-c) has been added to mem-monitor-smaps. Manual page has
    been updated to include these changes.

sp-memusage (1.0.9), 17 Apr 2008:
  * mem-monitor-smaps: Private clean information has been added. 

sp-memusage (1.0.8), 15 Apr 2008:
  * Mem-monitor-smaps shows now memory usage changes in addition to totals.
  * Argument checking improvements in mem-monitor-smaps

sp-memusage (1.0.7), 20 Mar 2008:
  * New helper script (run-with-memusage) has been added.
  * Made naming of existing helper scripts more consistent:
    memusage -> mem- monitor,
    proc-memusage -> mem-monitor-proc,
    sum-smaps-private -> mem-smaps-private,
    sum-dirty-code-pages -> mem-dirty-code-pages

sp-memusage (1.0.6), 3 Mar 2008:
  * New tool (proc-memusage) and correspoding documentation have been
    added. 

sp-memusage (1.0.5), 6 Feb 2008:
  * Fixed SONAME missing from libmalloc to satisfy Lintian.

sp-memusage (1.0.4), 6 Nov 2007:
  * Include run-with-mallinfo convenience script for using mallinfo in
    Scratchbox. 

sp-memusage (1.0.3), 4 Oct 2007:
  * Separated sp-memusage completely from sp-libleaks source package


src/mem-monitor.c
/* ========================================================================= *
 * File: mem-momitor.c, part of sp-memusage
 * 
 * Copyright (C) 2005-2009 by Nokia Corporation
 * 
 * Author: Leonid Moiseichuk <leonid.moiseichuk@nokia.com>
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * version 2 as published by the Free Software Foundation. 
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 *
 * Description:
 *    Read /proc/meminfo and return memory usage values at given interval:
 *    free, used, usage, total.
 *
 * History:
 *
 * 01-Jun-2009
 * - Moved some code to mem-monitor-util.{ch}, that is now shared with
 *   mem-monitor and mem-cpu-monitor.  Removed mem-monitor.h.
 * - Call fflush(stdout) after printing, to make sure that the lines get
 *   flushed out of C library buffers when output is redirected to file.
 *
 * 17-Sep-2008 Eero Tamminen
 * - Use term "avail" instead of "free" as although the shown amount
 *   of memory is available for re-use, it isn't tehnically free
 *   but mostly in use as buffers & caches
 * 
 * 24-Aug-2006 Eero Tamminen
 * - add -% to usage and ':' to end of each header name
 * 
 * 20-Jun-2006 Leonid Moiseichuk
 * - added flags for low/high memory notifications.
 * - added timestamps to the output, show numbers as KBs
 *
 * 28-Apr-2006 Leonid Moiseichuk
 * - removed slab information which makes users confused.
 *
 * 27-Sep-2005 Leonid Moiseichuk
 * - initial version created using memlimits.c as a prototype.
 *
 * ========================================================================= */

/* ========================================================================= *
 * Includes
 * ========================================================================= */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <ctype.h>

#include "mem-monitor-util.h"

/* ========================================================================= *
 * Definitions.
 * ========================================================================= */

/* Structure that used to report about memory consumption */
typedef struct
{
    size_t  total;     /* Total amount of memory in system: RAM + swap */
    size_t  free;      /* Free memory in system, bytes                 */
    size_t  used;      /* Used memory in system, bytes                 */
    size_t  util;      /* Memory utilization in percents               */
} MEMUSAGE;

/* Compile-time array capacity calculation */
#define CAPACITY(a)  (sizeof(a) / sizeof(*a))

/* Correct division of 2 unsigned values */
#define DIVIDE(a,b)  (((a) + ((b) >> 1)) / (b))

/* ------------------------------------------------------------------------- *
 * memusage -- returns memory usage for current system in MEMUSAGE structure.
 * parameters:
 *    usage - parameters to be updated.
 * returns:
 *    0 if values loaded successfuly OR negative error code.
 * ------------------------------------------------------------------------- */
static int memusage(MEMUSAGE* usage)
{
   /* Check the pointer validity first */
   if ( usage )
   {
      static MEMINFO vals[] =
      {
         { "MemTotal:",  0 },
         { "SwapTotal:", 0 },
         { "MemFree:",   0 },
         { "Buffers:",   0 },
         { "Cached:",    0 },
         { "SwapCached:",0 },
         { "SwapFree:",  0 }
      }; /* vals */

      /* Load values from the meminfo file */
      if ( CAPACITY(vals) == parse_proc_meminfo(vals, CAPACITY(vals)) )
      {
         /* Discover memory information using loaded numbers */
         usage->total = vals[0].value + vals[1].value;
         usage->free  = vals[2].value + vals[3].value + vals[4].value + vals[5].value + vals[6].value;
         usage->used  = usage->total - usage->free;
         usage->util  = DIVIDE(100 * usage->used, usage->total);

         /* We are succeed */
         return 0;
      }
      else
      {
         /* Clean-up values */
         memset(usage, 0, sizeof(usage));
      }
   }

   /* Something wrong, shows as error */
   return -1;
} /* memusage */

int main(const int argc, const char* argv[])
{
   /* Update interval once per 3 seconds by default */
   unsigned period = 3;
   
   if (argc > 2 || (argc == 2 && !isdigit(argv[1][0]))) 
   {
      fprintf(stderr, "\nusage: %s [output interval in secs]\n\n", *argv);
      exit(1);
   }
   else if (argc == 2)
   {
      period = strtoul(argv[1], NULL, 0);
   }

   /* We must print data always */
   if (nice(-19) == -1) {
      perror("Warning: failed to change process priority.");
   }

   printf ("time:\t\ttotal:\tavail:\tused:\tuse-%%:\tstatus:\n");
   while (1)
   {
       MEMUSAGE usage;

      /* Load all values from meminfo file */
      if (0 == memusage(&usage))
      {
         const time_t tv = time(NULL);
         struct tm*   ts = localtime(&tv);
         const char*  bg = (check_flag("/sys/kernel/low_watermark") ? "BgKill" : "");
         const char*  lm = (check_flag("/sys/kernel/high_watermark") ? ",LowMem" : "");

          printf ("%02u:%02u:%02u\t%u\t%u\t%u\t%u\t%s%s\n",
                     ts->tm_hour, ts->tm_min, ts->tm_sec,
                     usage.total, usage.free, usage.used, usage.util,
                     bg, lm
                  );

	  fflush(stdout);
      }
      else
      {
         printf ("unable to load values from /proc/meminfo file\n");
         return -1;
      }

      sleep(period);
   }

   /* That is all */
   return 0;
} /* main */

/* ========================================================================= *
 *                    No more code in file mem-monitor.c                     *
 * ========================================================================= */
